feat: add support to TMIcon #146

ultralabed opened the pull request:
Resolves: #146

![Screen Shot 2020-07-29 at 11 43 49 AM](https://user-images.githubusercontent.com/9026735/88828498-f12d6000-d190-11ea-9c52-360615506076.png)
![Screen Shot 2020-07-29 at 11 44 00 AM](https://user-images.githubusercontent.com/9026735/88828537-fbe7f500-d190-11ea-8421-f946692237b4.png)
![Screen Shot 2020-07-29 at 11 46 53 AM](https://user-images.githubusercontent.com/9026735/88828699-38b3ec00-d191-11ea-99a3-d5f40860c287.png)

ultralabed committed:
feat: add support to TMIcon #146

ultralabed committed:
Merge branch 'master' into feature/add-support-to-TMIcon

ultralabed committed:
fix: Update correct 3rd party npm registry

ultralabed committed:
fix: remove unused asset

ultralabed committed:
feat: use svgs instead of fonts for TMIcon

ultralabed committed:
Merge branch 'master' into feature/add-support-to-TMIcon

ultralabed committed:
feat: Update TMIcon to use icon instead of name property

ultralabed committed:
fix: remove unused rollup copy plugin

ultralabed commented on packages/react-styled-ui/src/TMIcon/tmIconSvgMap.js:
```diff
@@ -0,0 +1,2209 @@
+          ))}
+        </g>
+      ),
+      viewBox: '0 0 1060 1060',
```
@joshuatai Can you help to double confirm that the exact viewBox we can use? If we use '0 0 960 960' some of the icon's right and bottom edge will be cropped.

ultralabed replied:
Never mind, I will import icon with `0 0 16 16` and we can ignore this problem.

ultralabed replied:
Part of the icon will use `0 0 1060 1060` as a workaround, will be tackled in a separate PR.

ultralabed commented on packages/react-styled-ui/src/TMIcon/index.js:
```diff
@@ -0,0 +1,27 @@
+  return (
+    <Icon
+      ref={ref}
+      icon={`_core.tmicon-${icon}`}
```
@cheton Let me know if we should remove the `_core` prefix here. If yes, I will make new changes accordingly.

cheton replied:
yes! please remove it if possible

ultralabed replied:
Will do.

ultralabed commented on packages/styled-ui-docs/components/CodeBlock.jsx:
```diff
@@ -10,7 +10,7 @@ import * as ReactTable from 'react-table';
 import { codeBlockLight, codeBlockDark } from '../prism-themes/styled-ui';
 import FontAwesomeIcon from './FontAwesomeIcon';
-import TMIcon from './TMIcon';
+import iconList from './tmicon-list';
```
I am wondering if we can reuse `tmiconSvgMap` so that we don't have to maintain two similar copies of icon source. The downside of reusing `tmiconSvgMap` is that we will bundle it into `react-styled-ui` in order to use it in doc.

cheton replied:
你可以把 tmIconMap 和 tmIconSvgMap 分成兩個不同的檔案

* tmIconMap.js -> icon list
* tmIconSvgMap.js -> used to compose React components

然後在 docs 用 relative import

```js
import tmIconMap from '../../react-styled-ui/src/TMIcon/tmIconMap';
```

cheton replied:
@austinhung81 @ultralabed @joshuatai 

以後也許可以這樣，把那些 icon list 放在新的 package

Example:

```
packages/tonic-ui-icons
```

這樣就可以 share 同一份 source 了，然後 `tonic-ui-icons` 也有自己的 build script

cheton replied:
@joshuatai @austinhung81 @tinaClin @roth1002 @freewill0918 

1.0 之後如果這些都叫做 Tonic Icon，那應該是直接用 `<Icon />` 就好，不需要多一個 `<TMIcon />` ?

roth1002 replied:
就叫 ```<Icon />``` 吧

ultralabed replied:
Well noted. Changes below will be made.
1. Create `tmIconMap.js` to host icon configurations and details.
2. Create `tmIconSvgMap.js` to compose React SVG components.
3. Rename `TMIcon` to `Icon`.

ultralabed committed:
fix: Address PR comments

1. Remove _core prefix for internal icons
2. Create tmIconMap to host icon list
3. Create tmIconSvgMap to host composed React component
4. Rename TMIcon to Icon
5. Add szie/color example to Icon doc

ultralabed committed:
fix: disable no-array-index-key since there is nothing unique about them

ultralabed committed:
Merge branch 'master' into feature/add-support-to-TMIcon

ultralabed committed:
fix: update outdated TMIcon usage in pagination doc

ultralabed committed:
fix: fix Threat Type svg path incorrect
